contribute/design/1.x/ndm/20190905-ndm-exporter-integration.md

```diff
@@ -0,0 +1,180 @@
+Also the metrics should be available on demand at an end point from where users
+can query and get relevant information. This will help in monitoring of the block
+devices rather at the pool or volume level.
+
+### Goals
+
+- A cluster level exporter that exposes static and rarely changing metrics of
+  block devices like device state and UUIDs
+- A node level exporter that exposes continously varying metrics like temperature
+  and free space on the block device
+
+### Non-Goals
+
+- The exporter will be working only at the block level and will not provide
+  metrics about pool or volumes that are built on top of these devices.
+
+## Proposal
+
+### User Stories
+
+#### Rarely changing metrics
+For a high level view of storage on the cluster, data points like when the
+disk went offline, total capacity of the disk etc should be available to the
+cluster admin.
+
+#### Frequently changing metrics
+For deep monitoring of the storage devices, metrics like IOPS, drive temperature etc
+are required and these data should be queried from each block device from each node.
+
+### Implementation Details/Notes/Constraints
+
+The implementation details of exporter for collecting block level metrics. The current
+implementation just stores all the data related to the block devices in etcd, which
+is difficult to retrieve by monitoring systems like prometheus.
+
+#### Current Implementation
+In the current implementation all the details related to the block devices are stored
+in etcd. Even information like temperature is stored in the custom resource. The data
+in etcd is updated only when a udev event happens on the system.
+
+#### Shortcomings of current implementation
+Since the time at which udev events occur cannot be predicted and is not so frequent
+the metric data stored in etcd will be obselete most of the time. Also, this approach
+cannot be used in cases of continously varying metrics like IOPS and free space on the
+drive.
+
+#### Proposed Implementation
+
+There will be 2 components to the exporter. One will be running at the cluster level
+and the other at node level. The components can be customized depending on the metrics
+you need to fetch from the cluster.
+
+##### Cluster level Exporter
+This component of the exporter will run at the cluster level and there will be only one
+running instance of this component at any time in the cluster. The primary responsibility
+of this exporter is to take static / rarely changing data about the block devices from
+etcd and expose it as prometheus metrics. The static data like Model number, UUID etc
+will be cached in the exporter and metrics like state(Offline/Online) of the device will be
+fetched from etcd everytime a request comes at the end point for the metrics
+
+##### Node level Exporter
+The node level exporter runs as a daemon in each node and collects all the metrics of the
+block device like IOPS, temperature, free space etc. It exposes a rest endpoint which can
```

Comment:
free space of the disk or the file system created on the disk? What is the meaning of free space of the blockdevice?

Reply:
These are a sample list of metrics that can be exposed. The actual metrics depends on the collector used to fetch them. Currently free space metrics is not implemented.

I think free space can be measured only if a recognisable filesystem is present on the disk. Therefore the available space will be that of the filesystem.